# HPE Container Platform API series  - Part 3
## Launching cloud-native stateless applications and non-cloud-native stateful applications, programmatically through REST API calls, as a Kubernetes Tenant user.


**Requirements:**
- HPE Container Platform deployment
- IP address of FQDN of the HPE Container Platform's controller host
- a Kubernetes tenant user account   

**Utilities:**   
- cURL  
- Jupyter Notebook server with bash kernel installed
- kubectl, kubectl hpecp plug-in

**Definitions:**
- *HPE Container Platform* is an enterprise-grade container platform designed to deploy both cloud-native and non-cloud-native applications whether on-premises, at the edge, in multiple public clouds, or in a hybrid model. This makes the HPE Container Platform ideal for helping application developers and data scientists accelerate their application development and deployment on **containers**, on-demand through a self-service portal and a RESTful API that surfaces programmable access. To learn more about HPE Container Platform visit the [HPE DEV portal](https://developer.hpe.com/platform/hpe-container-platform/home) and check out the blog articles.

- *tenant:* A tenant is a group of users created by the platform administrator. A tenant can represent for example, an office location, a business unit, an organization, a project, an application. A tenant is allocated a set of resources (CPU/GPU, RAM, Storage, App Store images, Kubernetes cluster) by the platform administrator. All the resources used by a tenant are not shared with other tenants. A tenant user is granted the role of member or admin for the tenant.


HPE Container Platform provides two types of tenants:  
* Big data, AI/ML tenants that exist within the context of HPE Container Platform.
* Kubernetes tenants that exist within the context of one or more Kubernetes clusters managed by HPE Container Platform.
  
Here, in this part 3 of the series, I will cover how Kubernetes tenant users can deploy, using REST API, both cloud-native stateless, microservices applications, and non-cloud-native distributed stateful AI/analytics **KubeDirector** applications on Kubernetes cluster managed by HPE Container Platform. Tenant users will then use kubectl to interact directly with the Kubernetes cluster in the context of their tenant.

KubeDirector, also known as Kubernetes Director, is an open source project led by HPE that enables the running of stateful analytics workloads on Kubernetes. 

## The HPE Container Platform API Reference
The HPE Container Platform REST API allows you to achieve multiple actions programmatically, from performing administrative tasks, deploying cloud-native and non-cloud-native applications to scoring trained Machine Learning models.

Before you can call the HPE Container Platform API, you need to know what calls can be placed. The REST API reference documentation describes each object type, along with the supported operations, request input parameters, response model and response codes. 
To access the REST API reference documentation, obtain the IP address or hostname of the current active HPE Container Platform controller host from the administrator of the platform. Then in a web browser, navigate to the following URL:

``` 
http(s)://[Controller-IP-address-name]/apidocs
```

All the REST API calls are in the form: 
``` 
An HTTP VERB such as (GET , POST, DELETE, PUT, PATCH, UPDATE),  
A target API object: http(s)://[Controller-IP-address]:8080/api/v2/[object]
```

## Session Authentication in a multi-tenant environment
With the exception of some API calls, most of the REST API calls you can do against the HPE Container Platform API requires authentication. HPE Container Platform uses a *‘session location’* to use as operation context. In a multi-tenant environment, you request an authentication session location by issuing an authentication request in the following form:
* Call the API to request a new login session, providing username/password credentials as well as the Tenant name in the JSON body.  The user must be a valid tenant user credentials with a role (member or admin) in the requested tenant. 
* Extract the resource path of the created *session location* object from the JSON response header,
* For each subsequent call, set a new HTTP Header with its key set to *X-BDS-SESSION* and its value set to the session location value and used as the *working tenant* context. 

``` 
Note: the session location will expire after 24 hours. 
```   

If you are not already familiar with REST API calls, I encourage you to check-out the [Understanding API basics and the value they provide](https://developer.hpe.com/blog/understanding-api-basics-and-the-value-they-provide) blog on HPE Developer Community portal. It explains you REST API concepts such as HTTP verbs you call against the API, the headers, and payloads used when making API calls.  

**cURL:** You will use cURL to make API requests. Information on cURL can be found [here](https://curl.haxx.se/)

#### Initialize the environment.

**IMPORTANT: Before running the next cells, please make sure to adjust the environment variables below according to your Student ID, tenant username and password.**

In [1]:
#
# environment variables to be adjusted by the student
#
###student="studentXX" # your Jupyter Notebook student Identifier (i.e.: student<xx>)
###username="studentXX" # your tenant login credentials - username and password (it matches your Notebook Student account)
###password="Thepassword"
student="test"
username="youruser"
password="yourpassword"
#
# fixed environment variables setup by the HPE CP administrator
#
controller_endpoint="controller.hpedevlab.net:8080"
controller_host="controller.hpedevlab.net"
tenantname="K8sHackTenant"
k8sClusterId="1"  #this is the K8s Cluster Id provided by the HPE CP admisnistrator and assigned to your K8s tenant.
helloWorldApp="hello-world-app.yaml" # the application manifest you will deploy in this lab
tensorFlowApp="tensorflow-notebook-config-cluster.yaml" # the kubedirector application cluster configuration

#### Authenticate as Tenant user in the specified tenant:

In [2]:
sessionlocation=$(curl -k -i -s --request POST "https://${controller_endpoint}/api/v2/session" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--data-raw '{
"name": "'"$username"'",
"password": "'"$password"'",
"tenant_name": "'"$tenantname"'"
}' | grep Location | awk '{print $2}' | tr -d '\r') #we remove any cr that might exist
echo "This is your session location: " $sessionlocation
SessionId=$(echo $sessionlocation | cut -d'/' -f 5) # extract sessionId for later, for logout
echo "This is your session_Id:" $SessionId

This is your session location:  /api/v2/session/050eda34-4807-4790-8396-13ff2521217c
This is your session_Id: 050eda34-4807-4790-8396-13ff2521217c


#### Make a quick check to ensure you can make REST API calls within your tenant working context.

In [3]:
curl -k -s --request GET "https://${controller_endpoint}/api/v2/session" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' | jq  #using jq to pretty print the JSON reponse of the API call 

{
  "_links": {
    "self": {
      "href": "/api/v2/session"
    }
  },
  "session_expiry_minutes": 1440,
  "_embedded": {
    "sessions": [
      {
        "_links": {
          "self": {
            "href": "/api/v2/session/050eda34-4807-4790-8396-13ff2521217c"
          },
          "all_sessions": {
            "href": "/api/v2/session"
          }
        },
        "user": "/api/v1/user/60",
        "user_name": "user",
        "tenant": "/api/v1/tenant/5",
        "tenant_name": "K8sHackTenant",
        "role": "/api/v1/role/3",
        "role_name": "Member",
        "expiry": "2020-4-7 23:57:52",
        "expiry_time": 1586296672,
        "is_site_admin_view": false,
        "is_cluster_superuser": false
      }
    ]
  }
}


## Deploying a simple stateless application

#### -1- Deploy Hello World containerized application
You will deploy a simple Hello World application on the Kubernetes cluster made available to your tenant. The REST API call requires the **kubectl** operation (create, apply, delete) and the application manifest (YAML file that describes the attributes of the application) in a base64 encoded form.  

As you are all sharing the same tenant context and Kubernetes cluster resources, let's make sure your application deployment name will be unique among the tenant users. Here we replace the string "test" with your "username".

In [4]:
sed -i "s/example/${username}/g" $helloWorldApp
cat $helloWorldApp

apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-world-user
spec:
  selector:
    matchLabels:
      run: load-balancer-user
  replicas: 2
  template:
    metadata:
      labels:
        run: load-balancer-user
    spec:
      containers:
        - name: hello-world-user
          image: gcr.io/google-samples/node-hello:1.0
          ports:
            - containerPort: 8080
              protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  name: hello-world-service-user
spec:
  selector:
    run: load-balancer-user
  ports:
  - name: http-hello
    protocol: TCP
    port: 8080
    targetPort: 8080
  type: NodePort


In [5]:
#encode the application description file in base64
myapp=$(base64 $helloWorldApp)
#echo $myapp

In [6]:
curl -k -s --request POST "https://${controller_endpoint}/api/v2/k8scluster/${k8sClusterId}/kubectl" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--data-raw '{
"data": "'"$myapp"'",
"op": "apply"
}'

deployment.apps/hello-world-user created
service/hello-world-service-user created

After a minute or so, your application should be deployed on the Kubernetes cluster.   
You will directly interact with the Kubernetes cluster using **kubectl** (the command line configuration tool for Kubernetes).  
Let's get the service endpoints of your application using kubectl as well as the kubectl hpecp plug-in.

#### -2- Establish the kubectl authenticated context 

The kubectl hpecp plug-in is an extension of the standard kubectl command line. The plugin handles HPECP login and session token management. It is used to establish **HPECP-authenticated** kubectl requests to HPECP-managed Kubernetes services.   

Let's get the kubeconfig file for your working tenant context. The kubeconfig file contains kubernetes context that tenant users can interact with, based on their assigned role (tenant admin or tenant member).

In [7]:
#kubectl version --client
#kubectl hpecp version
kubectl hpecp refresh --insecure-skip-tls-verify ${controller_host} --hpecp-user="${username}" --hpecp-pass="${password}"



Retrieved new Kube Config from HPECP server at controller.hpedevlab.net:8080.
The KUBECONFIG environment variable HAS NOT been set.
Your current session WILL NOT have the new configuration.
To persist these changes by loading all current Kube Config
values into your default Kube Config file, run the
following command:

    KUBECONFIG="/home/test/.kube/.hpecp/controller.hpedevlab.net/config:/home/test/.kube/config-backup" kubectl config view --raw > /home/denis/.kube/config

To persist these changes by changing your local KUBECONFIG
environment variable, run the following command:

    export KUBECONFIG="/home/test/.kube/.hpecp/controller.hpedevlab.net/config"

CAUTION - both of these commands will OVERWRITE your current
Kube Config settings. This is probably what you want, but
to confirm that this command will not break your system,
run the following command to view the resulting Kube
Config file:

    KUBECONFIG="/home/test/.kube/.hpecp/controller.hpedevlab.net/config:/home/test/.ku

In [8]:
#define the Kubeconfig file as a shell environment variable
export KUBECONFIG="/home/${student}/.kube/.hpecp/controller.hpedevlab.net/config"
#echo $KUBECONFIG

Check your working tenant context

In [9]:
kubectl config view

apiVersion: v1
clusters:
- cluster:
    insecure-skip-tls-verify: true
    server: https://gateway1.etc.fr.comm.hpecorp.net:9500
  name: k8scluster1
contexts:
- context:
    cluster: k8scluster1
    namespace: k8shacktenant
    user: HPECP-user
  name: k8scluster1-K8sHackTenant-user
current-context: k8scluster1-K8sHackTenant-user
kind: Config
preferences: {}
users:
- name: HPECP-user
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - hpecp
      - authenticate
      - gateway1.etc.fr.comm.hpecorp.net:8080
      - --hpecp-user=user
      - --hpecp-token=/api/v2/session/e28538a9-f794-48e0-a3b5-1a42a817d660
      - --hpecp-token-expiry=1586296791
      - --force-reauth=false
      - --insecure-skip-tls-verify=true
      command: kubectl
      env: null


#### -3- Get the POD and Services for your deployed application:

In [10]:
kubectl get pod,service | grep ${username}

pod/hello-world-user-7d79c6c7dc-5vnrd   1/1     Running   0          2m16s
pod/hello-world-user-7d79c6c7dc-cd757   1/1     Running   0          2m16s
service/hello-world-service-user   NodePort   10.96.30.31   <none>        8080:31130/TCP   2m16s


In [11]:
kubectl describe service hello-world-service-${username}

Name:                     hello-world-service-user
Namespace:                k8shacktenant
Labels:                   hpecp.hpe.com/hpecp-internal-gateway=true
Annotations:              hpecp-internal-gateway/8080: gateway1.etc.fr.comm.hpecorp.net:10002
                          kubectl.kubernetes.io/last-applied-configuration:
                            {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"name":"hello-world-service-user","namespace":"k8shacktenant"},"spec"...
Selector:                 run=load-balancer-user
Type:                     NodePort
IP:                       10.96.30.31
Port:                     http-hello  8080/TCP
TargetPort:               8080/TCP
NodePort:                 http-hello  31130/TCP
Endpoints:                10.192.0.167:8080,10.192.1.39:8080
Session Affinity:         None
External Traffic Policy:  Cluster
Events:
  Type    Reason  Age    From         Message
  ----    ------  ----   ----         -------
  Normal  HpeCp   2m31s  hp

#### -4- Get the service endpoints
HPE Container Platform will automatically maps the NodePort Service endpoint to the HPE CP gateway (proxy) host.

In [12]:
myservice="hello-world-service-"${username}
appURL=$(kubectl describe service/"${myservice}" | grep gateway1 | awk '{print $3}')
myapp_endpoint="https://${appURL}"
echo "Your application service endpoint is: "$myapp_endpoint

Your application service endpoint is: https://gateway1.etc.fr.comm.hpecorp.net:10002


#### -5- Check your application is responding:

In [13]:
curl -k -s "${myapp_endpoint}"

Hello Kubernetes!

#### -6- Go through some cleanup
Delete your application deployment and services. After a minute or so, you should get the message: deployment deleted and service deleted.

In [14]:
curl -k -s --request POST "https://${controller_endpoint}/api/v2/k8scluster/${k8sClusterId}/kubectl" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--data-raw '{
"data": "'"$myapp"'",
"op": "delete"
}'

deployment.apps "hello-world-user" deleted
service "hello-world-service-user" deleted

## Deploying a KubeDirector Application

HPE has been working within the open source Kubernetes community to add capabilities that enable the running of stateful analytics workloads on Kubernetes.  The open source project is known as **Kubernetes Director** or **KubeDirector** for short.   
With KubeDirector, enterprises can deploy all their enterprise applications on a common Kubernetes framework.
KubeDirector is deployed as a custom controller (aka an operator) by default on every Kubernetes cluster managed by the HPE Container Platform.

For more information about KubeDirector, check out the HPE DEV portal [here](https://developer.hpe.com/platform/hpe-container-platform/home).

#### -1- List the installed KubeDirector Application
The Kubernetes cluster managed by HPE Controller platform has three (3) pre-configured KubeDirector application types installed out of the box.

<img src="Pictures/HPECP-KubeDirectorApp-GUI.png" height="800" width="600" align="left">

In [15]:
kubectl get kubedirectorapp

NAME                  AGE
centos7x              13d
ml-jupyter-notebook   13d
spark221e2            13d


A KubeDirectorApp describes an application type, its images, service endpoints, and if appropriate, the root file system directories (e.g.: /etc, /bin, /opt, /var, /usr) of the containers to persist.

In [17]:
kubectl describe kubedirectorapp ml-jupyter-notebook 

Name:         ml-jupyter-notebook
Namespace:    k8shacktenant
Labels:       <none>
Annotations:  token:
                execute this command to get the authentication token 'kubectl exec <pod-name> jupyter notebook list' and use this token in Jupyter noteboo...
API Version:  kubedirector.hpe.com/v1beta1
Kind:         KubeDirectorApp
Metadata:
  Creation Timestamp:  2020-03-24T08:25:15Z
  Generation:          2
  Resource Version:    2874116
  Self Link:           /apis/kubedirector.hpe.com/v1beta1/namespaces/k8shacktenant/kubedirectorapps/ml-jupyter-notebook
  UID:                 cb7a6132-f321-45d6-81f5-635f0b79e61e
Spec:
  Config:
    Role Services:
      Role ID:  controller
      Service I Ds:
        jupyter-nb
    Selected Roles:
      controller
  Config Schema Version:   7
  Default Config Package:  <nil>
  Distro ID:               bluedata/tensorflow
  Label:
    Description:  TensorFlow GPU with Jupyter notebook
    Name:         TensorFlow + Jupyter
  Roles:
    Cardinality:

#### -2- Deploy the TensorFlow with Jupyter Notebook stateful virtual application cluster
Similar to any Kubernetes based workload deployment, a configuration manifest YAML file is used to create virtual application clusters that instantiate the defined KubeDirector App type.

We need to make sure the instance of the kubedirector application is unique among your tenant. We also need to convert the application configuration manifest in a base64 encoded form.

In [18]:
sed -i "s/example/${username}/g" $tensorFlowApp
cat $tensorFlowApp

---
apiVersion: "kubedirector.hpe.com/v1beta1"
kind: "KubeDirectorCluster"
metadata: 
  name: "ml-jupyter-notebook-user"
spec: 
  app: "ml-jupyter-notebook"
  appCatalog: "local"
  roles: 
    - 
      id: "controller"
      members: 1
      resources: 
        requests: 
          memory: "4Gi"
          cpu: "2"
          nvidia.com/gpu: "0"
        limits: 
          memory: "4Gi"
          cpu: "2"
          nvidia.com/gpu: "0"


The Configuration file specifies the application instance name, the number of nodes, their roles and the compute size.
Let's deploy it now.

In [19]:
mytfapp=$(base64 $tensorFlowApp)
#echo $mytfapp
curl -k -s --request POST "https://${controller_endpoint}/api/v2/k8scluster/${k8sClusterId}/kubectl" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--data-raw '{
"data": "'"$mytfapp"'",
"op": "apply"
}'

kubedirectorcluster.kubedirector.hpe.com/ml-jupyter-notebook-demouser created

After a minute or so, you should get the message kubedirectorcluster/Your-instance-name created.

#### -3- Inspecting the deployed virtual application cluster 
The virtual cluster will be represented in the Kubernetes cluster by a resource of type **KubeDirectorCluster**, with the name that was indicated inside the YAML file used to create it. A kubeDirectorCluster resource is an instantiation of a KubeDirector App.

In [49]:
clusterName="ml-jupyter-notebook-${username}"
kubectl get kubedirectorcluster $clusterName

NAME                           AGE
ml-jupyter-notebook-user   30m


After creating the instance of the KubeDirector App, you can use kubectl command below to observe its status and the Kubernetes resources that make up the virtual cluster (statefulsets, pods, services, persistent volume claim if any) and any events logged against it.

The virtual cluster status indicates its overall "state" (top-level property of the status object). It should have a value of **"configured"**. The first time a virtual cluster of a given app type is created, it may take some minutes to reach "configured" state, as the relevant Docker image must be downloaded and imported.

In [50]:
kubectl describe kubedirectorcluster $clusterName

Name:         ml-jupyter-notebook-user
Namespace:    k8shacktenant
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"kubedirector.hpe.com/v1beta1","kind":"KubeDirectorCluster","metadata":{"annotations":{},"name":"ml-jupyter-notebook-use...
API Version:  kubedirector.hpe.com/v1beta1
Kind:         KubeDirectorCluster
Metadata:
  Creation Timestamp:  2020-04-06T22:08:29Z
  Finalizers:
    kubedirector.hpe.com/cleanup
  Generation:        1
  Resource Version:  3782690
  Self Link:         /apis/kubedirector.hpe.com/v1beta1/namespaces/k8shacktenant/kubedirectorclusters/ml-jupyter-notebook-user
  UID:               9a970b42-8fe2-42c1-b8bd-9e97faea0840
Spec:
  App:          ml-jupyter-notebook
  App Catalog:  local
  Roles:
    Id:       controller
    Members:  1
    Resources:
      Limits:
        Cpu:             2
        Memory:          4Gi
        nvidia.com/gpu:  0
      Requests:
        Cpu:             2
        Me

In [51]:
kubectl get all -l kubedirector.hpe.com/kdcluster=$clusterName

NAME               READY   STATUS    RESTARTS   AGE
pod/kdss-kv955-0   1/1     Running   0          30m

NAME                     TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
service/kdhs-ktrzh       ClusterIP   None           <none>        8888/TCP         30m
service/s-kdss-kv955-0   NodePort    10.96.194.58   <none>        8888:30124/TCP   30m

NAME                          READY   AGE
statefulset.apps/kdss-kv955   1/1     30m


Your application cluster is made up of a statefulset, a pod and a service.

#### -4- Get the service endpoints
To get a report on all services related to a specific virtual cluster, you can use a form of "kubectl get" that matches against a value of the **kubedirector.hpe.com/kdcluster=YourClusterApplicationName** label.

In [70]:
tfappURL=$(kubectl describe service -l  kubedirector.hpe.com/kdcluster=${clusterName} | grep gateway1 | awk '{print $3}')
mytfapp_endpoint="https://${tfappURL}"
echo "Your application virtual cluster service endpoint is: "$mytfapp_endpoint

Your application virtual cluster service endpoint is: https://gateway1.etc.fr.comm.hpecorp.net:10002


#### -5- Check the application service is responding

In [53]:
curl -k -L -s -i "${mytfapp_endpoint}" | grep "HTTP/1.1"

HTTP/1.1 302 Found
HTTP/1.1 302 Found
HTTP/1.1 200 OK


#### -6- Fetch the authentication token of TensorFlow Jupyter Notebook 

Execute this command to get the authentication token 'kubectl exec pod-name jupyter notebook list' and use this token in Jupyter notebook

In [71]:
mytfapp_pod=$(kubectl get pod -l kubedirector.hpe.com/kdcluster=${clusterName} | grep kdss | awk '{print $1}') 
#echo $mytfapp_pod
kubectl exec ${mytfapp_pod} jupyter notebook list

Currently running servers:
http://0.0.0.0:8888/?token=110189cfab23d4b75b840d5a45a93a9472fd1992fdc72447 :: /tf


#### -7- Connect to your application cluster

You can connect to your tensorflow framework with Jupyter Notebook and start build your ML model (for example a regression model), train the model and make prediction.

Open a new tab in your browser, and connect to the service endpoint: https://77.158.163.130:YourPortNumber?token=YourToken
where 77.158.163.130 is the proxy gateway, port number is the port you get for your service endpoint, and the token is the authentication token you have just fetched. 

![TensorFlow Jupyter Notebook](Pictures/kubedirector-tf-jupyterNotebook.png)

#### -8- Go through some cleanup

* First delete the kubedirector application cluster instance

In [72]:
curl -k -s --request POST "https://${controller_endpoint}/api/v2/k8scluster/${k8sClusterId}/kubectl" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--data-raw '{
"data": "'"$mytfapp"'",
"op": "delete"
}'

kubedirectorcluster.kubedirector.hpe.com "ml-jupyter-notebook-demouser" deleted

* Although session have a time to live (TTL) of 24 hours, it is best practice in REST API programming to cleanup and delete those sessions when done. We can use a DELETE /api/v2/session/SessionId to achieve this.

In [73]:
curl -k -i -s --request DELETE "https://${controller_endpoint}/api/v2/session/${SessionId}" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json'

HTTP/1.1 204 No Content
Access-Control-Allow-Origin: *
Content-Length: 0
Content-Type: text/plain
Date: Mon, 06 Apr 2020 23:01:15 GMT
Server: BlueData EPIC 5.0



The status *204 No Content* means the session has been deleted.

* Finally, reset your applications YAML files

In [74]:
#reset the application deployment name in the YAML file
sed -i "s/${username}/example/g" $helloWorldApp
sed -i "s/${username}/example/g" $tensorFlowApp
cat $helloWorldApp
cat $tensorFlowApp

apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-world-example
spec:
  selector:
    matchLabels:
      run: load-balancer-example
  replicas: 2
  template:
    metadata:
      labels:
        run: load-balancer-example
    spec:
      containers:
        - name: hello-world-example
          image: gcr.io/google-samples/node-hello:1.0
          ports:
            - containerPort: 8080
              protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  name: hello-world-service-example
spec:
  selector:
    run: load-balancer-example
  ports:
  - name: http-hello
    protocol: TCP
    port: 8080
    targetPort: 8080
  type: NodePort
---
apiVersion: "kubedirector.hpe.com/v1beta1"
kind: "KubeDirectorCluster"
metadata: 
  name: "ml-jupyter-notebook-example"
spec: 
  app: "ml-jupyter-notebook"
  appCatalog: "local"
  roles: 
    - 
      id: "controller"
      members: 1
      resources: 
        requests: 
          memory: "4Gi"
          cpu: "2"
          nvidia.com/gp

## Summary

In this tutorial, you learned how Kubernetes tenant users can deploy, using REST API, both cloud-native stateless, microservices applications, and non-cloud-native distributed stateful AI/analytics KubeDirector applications on Kubernetes clusters managed by HPE Container Platform. You also used the standard Kubernetes command line (kubectl) as well as the HPE CP plugin to establish HPECP-authenticated kubectl requests in the context of your tenant user. 